library/modules/Units.cpp

```diff
@@ -122,7 +122,11 @@ constexpr uint32_t exclude_flags2 = (
 
 bool Units::isActive(df::unit *unit) {
     CHECK_NULL_POINTER(unit);
-    return !unit->flags1.bits.inactive;
+    if (unit->flags1.bits.inactive)
+        return false;
+    else if (unit->flags1.bits.move_state || unit->flags1.bits.can_swap)
+        return true; // These are always unset when leaving the map
+    return linear_index(world->units.active, &df::unit::id, unit->id) >= 0;
```

Comment:
`Units::isActive` is used in many many places. I'm worried that adding an O(N) search for every call will be a performance killer. I'm also concerned that changing the behavior of this function will break existing callers, both Lua and C++. I'd like to see some research done on those two points before merging this PR.

Reply:
I did a check on my fort, and all of my units in `units.active` had at least one of `inactive`, `move_state`, or `can_swap` set. There was an `inactive` unit without the other two flags set that had its `pos` directly above my well. It's possible the combination could represent a ghost or climbing unit? Probably not going to be a performance consideration given how infrequently those occur. Then there's the actual raiding units, which you won't hit while iterating `units.active`, and for which the check is mandatory when iterating `units.all`.

I went through all the uses of `isActive`, `getPosition`, `getUnitsInBox`, and `isUnitInBox` I could find, as well as the `inactive` flag. There's no change in behavior to anything iterating the `units.active` vector (or indirectly through `forCitizens`/`getCitizens`). (`isDead` is used instead of `isActive` for the case of actual dead units.) Anything iterating `units.all` wasn't properly considering raiding units, which was just `RemoteTools.cpp`.

I don't fully understand `plugins/preserve-rooms.cpp` but it didn't look like the changes break anything. You'd know better than me.